url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/processing/ComputationalGeometry.java

```diff
@@ -0,0 +1,314 @@
+    strategy proposed in [BAER2002].  For a given triangle and with respect
+    to triangle's containing plane, a poing lies in one of 7 regions:
+       - R1: inside triangle
+       - R2: outside triangle, near edge 1
+       - R3: outside triangle, near edge 2
+       - R4: outside triangle, near edge 3
+       - R5: outside triangle, near vertex 1
+       - R6: outside triangle, near vertex 2
+       - R7: outside triangle, near vertex 3
+    */
+    public static int triangleContainmentTest(
+        Vector3D p0, Vector3D p1, Vector3D p2, Vector3D p,
+        double distanceTolerance)
+    {
+        Vector3D n, a, b;
+
+        a = p1.substract(p0);
+        b = p2.substract(p0);
+        n = a.crossProduct(b);
+        n.normalize();
+
+        workPlane.setFromPointNormal(p0, n);
+
+        if ( workPlane.doContainmentTest(p, distanceTolerance) == 
+             Geometry.LIMIT ) {
+            // Barycentric coordinates test containment technique (Region 1)
+            Vector3D c = p.substract(p0);
+            double dot00, dot01, dot02, dot11, dot12, invDenom, u, v;
+            dot00 = a.dotProduct(a);
+            dot01 = a.dotProduct(b);
+            dot02 = a.dotProduct(c);
+            dot11 = b.dotProduct(b);
+            dot12 = b.dotProduct(c);
+            invDenom = 1 / (dot00 * dot11 - dot01 * dot01);
+            u = (dot11 * dot02 - dot01 * dot12) * invDenom;
+            v = (dot00 * dot12 - dot01 * dot02) * invDenom;
+            if ( (u >= 0) && (v >= 0) && (u + v <= 1) ) {
+                // R1
+                return Geometry.LIMIT;
+            }
+            else if ( (u <= 0) && (v >= 0) && (u + v <= 1) ) {
+                // R2
+                return
+                    lineSegmentContainmentTest(p0, p2, p, distanceTolerance);
+            }
+            else if ( (u >= 0) && (v <= 0) && (u + v <= 1) ) {
+                // R3
+                return
+                    lineSegmentContainmentTest(p0, p1, p, distanceTolerance);
+            }
+            else if ( (u >= 0) && (v >= 0) && (u + v >= 1) ) {
+                // R4
+                return
+                    lineSegmentContainmentTest(p1, p2, p, distanceTolerance);
+            }
+            else if ( (u <= 0) && (v <= 0) ) {
+                // R5
+                if ( VSDK.vectorDistance(p, p0) < distanceTolerance ) {
+                    return Geometry.LIMIT;
+                }
+            }
+            else if ( (u <= 0) && (v >= 1) ) {
+                // R6
+                if ( VSDK.vectorDistance(p, p2) < distanceTolerance ) {
+                    return Geometry.LIMIT;
+                }
+            }
+            else {
+                // R7
+                if ( VSDK.vectorDistance(p, p1) < distanceTolerance ) {
+                    return Geometry.LIMIT;
+                }
+            }
+        }
+        return Geometry.OUTSIDE;
+    }
+
+    public static void triangleMinMax(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 293-308 into a separate method.